swupd/create_manifests.go

```
@@ -0,0 +1,260 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package swupd
+
+import (
+	"errors"
+	"fmt"
+	//"io/ioutil"
+	"os"
+	"path/filepath"
+	"time"
+)
+
+// MinVersion indicates a minimum version build in which all content must be generated
+var MinVersion = false
+
+// UpdateInfo contains the meta information for the current update
+type UpdateInfo struct {
+	oldFormat   uint
+	format      uint
+	lastVersion uint32
+	version     uint32
+	bundles     []string
+	timeStamp   time.Time
+}
+
+func initBuildEnv(c config) error {
+	tmpDir := filepath.Join(c.stateDir, "temp")
+	// remove old directory
+	if err := os.RemoveAll(tmpDir); err != nil {
+		return err
+	}
+
+	// create new one
+	return os.Mkdir(tmpDir, os.ModePerm)
+}
+
+// initBuildDirs creates the following directory structure
+// StateDir/
+//    image/
+//        <version>/
+//            <bundle1>/
+//            ...
+//    LAST_VER
+func initBuildDirs(version uint32, groups []string, imageBase string) error {
+	verDir := filepath.Join(imageBase, fmt.Sprint(version))
+	for _, bundle := range groups {
+		if err := os.MkdirAll(filepath.Join(verDir, bundle), 0755); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func processBundles(ui UpdateInfo, c config) ([]*Manifest, error) {
+	newManifests := []*Manifest{}
+	for _, bundle := range ui.bundles {
+		oldM := &Manifest{}
+		oldMPath := filepath.Join(c.outputDir, fmt.Sprint(ui.lastVersion), "Manifest."+bundle)
+		if err := oldM.ReadManifestFromFile(oldMPath); err != nil {
+			// old manifest may not exist, continue
```

If we are ignoring the error, just ignore the result of ReadManifestFromFile.

Question: if the manifest exists but is malformed, should we still ignore it or should we fail processing?